src/crowdin_client.cpp

```diff
@@ -143,139 +158,331 @@ dispatch::future<void> CrowdinClient::Authenticate()
+                pathStr += '/';
+                pathStr += path.top();
+                path.pop();
+            }
+            if(!pathStr.empty())
+                i.pathName = str::to_wstring(pathStr) + i.pathName;
+        }
+
+        //TODO: get more until all branches gotten (if more than 500)    
+        return m_api->get(url + "/branches?limit=500");
+    }).then([this, url, prj](json r)
+    {
+        // Handle branches
+        std::map<int, std::string> branches;
+
+        for(const auto& i : r["data"])
+        {
+            const json& d = i["data"];
+            branches[d["id"]] = d["name"].get<std::string>();
+        }
+
+        for(auto& i : prj->files)
+            if(i.branchId != NO_ID)
+                i.pathName = "/" + str::to_wstring(branches[i.branchId]) + i.pathName;
+
+        return *prj;
+    });
 }
 
 
-dispatch::future<void> CrowdinClient::DownloadFile(const std::string& project_id,
-                                                   const std::wstring& file,
+dispatch::future<void> CrowdinClient::DownloadFile(const long project_id,
                                                    const Language& lang,
+                                                   const long file_id,
+                                                   const std::string& file_extension,
                                                    const std::wstring& output_file)
 {
-    // NB: "export_translated_only" means the translation is not filled with the source string
-    //     if there's no translation, i.e. what Poedit wants.
-    auto url = "/api/project/" + project_id + "/export-file"
-                   "?json="
-                   "&export_translated_only=1"
-                   "&language=" + lang.LanguageTag() +
-                   "&file=" + http_client::url_encode(file);
-    return m_api->download(url, output_file);
+    wxLogTrace("poedit.crowdin", "DownloadFile(project_id=%ld, lang=%s, file_id=%ld, file_extension=%s, output_file=%S)", project_id, lang.LanguageTag(), file_id, file_extension.c_str(), output_file.c_str());
+    wxString ext(file_extension);
+    ext.MakeLower();
+    return m_api->post(
+        "projects/" + std::to_string(project_id) + "/translations/builds/files/" + std::to_string(file_id),
+        json_data({
+            { "targetLanguageId", lang.LanguageTag() },
+            // for XLIFF and PO files should be exported "as is" so set to `false`
+            { "exportAsXliff", !(ext == "xliff" || ext == "xlf" || ext == "po") }
+        }))
+        .then([this, output_file] (json r) {
+            std::string url = r["data"]["url"];
+            wxURI uri(url);
+            // Per download (local) client must be created since different domain
+            // per request is not allowed by HTTP client backend on some platorms.
+            // (e.g. on Linux).
+            auto downloader = std::make_shared<crowdin_http_client>(*this, std::string((uri.GetScheme() + "://" + uri.GetServer()).mb_str()));
```

Comment:
Don't use `mb_str()`. Where you can, use `str_helpers.h` for conversions for clarify; here that would be `str::to_utf8()`.

`mb_str()` would do unexpected things, it is locale dependent.

Reply:
Addressed 9bc52ba2523397184c28c5f4184a00908ab103ae though noticed few other `mb_str()` in _your_ code

Reply:
> Addressed 9bc52ba though noticed few other mb_str() in your code .

I should have been more precise, sorry. The issue is no `mb_str()` as such, but using it with the default value for its argument, i.e. `wxConvLibc`, which performs the conversion using whatever the standard library sees as appropriate, which is related to the current locale. When we know the charset (here, ASCII, i.e. subset of UTF-8), it's best to be explicit. When possible, it's best to avoid relying on runtime-dependent charset.

I went to fix the other code right away (I truly appreciate such remarks!) , but  I see 4 uses of `mb_str` in the current tree. 3 of them use explicit conversion object (of which 1 deals with incorrectly encoded data). 1 is the bad implicit use, but it is when passing data to a function with const char* API, `u_setDataDirectory`, where it's the only option - that library is using libc locale.